### PR TITLE
Update uses of open ssl

### PR DESCRIPTION
**Pull request by adfoster-r7**
This PR:
- Updates uses of openssl objects to assign the key _after_ calling `.decrypt` / `.encrypt` method calls
- Runs `Lint/DeprecatedOpenSSLConstant` Rubocop rule

```
rubocop --only Lint/DeprecatedOpenSSLConstant -a modules lib
```

## Context

Ruby 3.0 ships with openssl 2.2.0 - which adds now throws an error if the key is set before calling decrypt/encrypt methods.
This PR fixes calls to OpenSSL to ensure methods are called in the right order.

For my machine:
```
2.6.5 :005 > OpenSSL::OPENSSL_LIBRARY_VERSION
 => "OpenSSL 1.1.1i  8 Dec 2020" 
2.6.5 :006 > OpenSSL::VERSION
 => "2.2.0" 
```

### Before

Running with Ruby 2.7 / OpenSSL 2.1.0:

```
docker run -it --rm -v $(pwd):$(pwd) -w $(pwd) ruby:2.7 /bin/bash
```

The following code works:

```ruby
$ irb

require 'openssl'
cipher = OpenSSL::Cipher.new("aes-128-cbc")
cipher.key = "A" * 16
cipher.encrypt
cipher.update("hello world") + cipher.final

=> "\\\x1D\xDF{\xCD+;\xBC\xCB\xDE+\x1E\xA7\xC9g\xF3"
```

### After

With Ruby 3.0 / OpenSSL 2.2.0:

```
docker run -it --rm -v $(pwd):$(pwd) -w $(pwd) ruby:3.0 /bin/bash
```

The following code now raises an exception:

```
Traceback (most recent call last):
        5: from /usr/local/bin/irb:23:in `<main>'
        4: from /usr/local/bin/irb:23:in `load'
        3: from /usr/local/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        2: from (irb):68:in `<main>'
        1: from (irb):68:in `update'
OpenSSL::Cipher::CipherError (key not set)
```

### Solution

The code needs updated to ensure the key is set _after_ the encrypt/decrypt call:

```ruby
require 'openssl'
cipher = OpenSSL::Cipher.new("aes-128-cbc")
cipher.encrypt
cipher.key = "A" * 16
cipher.update("hello world") + cipher.final
```

openssl 2.2.0 test example:
https://github.com/ruby/ruby/blame/42a16e5974c3919755013eeee0debb16ffb69ff5/test/openssl/test_cipher.rb#L308-L321

## Verification

I verified changes statically after updating the Gemfile to open ssl 2.2.0 and rerunning bundle:
```
gem 'openssl', '2.2.0'
```

I created small test snippets as sanity checks, which I've added as comments were appropriate

Remaining testing effort:
- [x] Test secrets_dump module. Passed, but noticed https://github.com/rapid7/metasploit-framework/issues/14832
- [x] Test meterpreter `hashdump`
- [x] `modules/post/windows/gather/smart_hashdump.rb`

### CODE REVIEW

File: lib/msf/core/post/windows/priv.rb [hunk 1]
```diff
@@ -306,6 +306,7 @@ def capture_lsa_key(bootkey)
       end
 
       rc4 = OpenSSL::Cipher.new("rc4")
+      rc4.decrypt
```
Comment (adfoster-r7) [inline lib/msf/core/post/windows/priv.rb:309]: This file was the only use of openssl that was being tested and impacted by the update

File: lib/msf/core/post/windows/priv.rb [hunk 2]
```diff
@@ -356,14 +357,15 @@ def decrypt_lsa_data(policy_secret, lsa_key)
...
+    aes.decrypt
     aes.key = sha256x.digest
 
     vprint_status("digest #{sha256x.digest.unpack("H*")[0]}")
 
     decrypted_data = ''
 
     (60...policy_secret.length).step(16) do |i|
-      aes.decrypt
+      aes.reset
```
Comment (adfoster-r7) [inline lib/msf/core/post/windows/priv.rb:368]: This file was the only use of openssl that was being tested and impacted by the update

File: modules/auxiliary/gather/windows_secrets_dump.rb [hunk 3]
```diff
@@ -407,17 +409,19 @@ def decrypt_user_hash(rid, hboot_key, enc_hash, pass, default)
...
+    d2.decrypt
     d2.padding = 0
     d2.key = des_k2
 
-    d1o = d1.decrypt.update(okey[0, 8])
+    d1o = d1.update(okey[0, 8])
     d1o << d1.final
 
-    d2o = d2.decrypt.update(okey[8, 8])
+    d2o = d2.update(okey[8, 8])
```
Comment (adfoster-r7) [inline modules/auxiliary/gather/windows_secrets_dump.rb:424]: Variations of this code snippet seemed to exist in a few places, maybe we're missing a shared module?

File: modules/post/multi/gather/lastpass_creds.rb [hunk 4]
```diff
@@ -726,13 +725,12 @@ def decrypt_vault_password(key, encrypted_data)
     return nil if key.blank? || encrypted_data.blank?
 
     if encrypted_data[0] == "!" # Apply CBC
-      decipher = OpenSSL::Cipher.new("AES-256-CBC")
+      decipher = OpenSSL::Cipher.new("AES-256-CBC").decrypt
       decipher.iv = encrypted_data[1, 16] # Discard !
```
Comment (adfoster-r7) [inline modules/post/multi/gather/lastpass_creds.rb:729]: Apparently setting the `iv` before calling `decrypt` is fine. This could be left as it is - but I decided for consistency to align with the other uses in the code base.

I didn't test the module, but wired up a code snippet to test it out instead

<details>

<summary>Code snippet</summary>

```
Verified with:
require 'openssl'

def decrypt_vault_password(key, encrypted_data)
  # return nil if key.blank? || encrypted_data.blank?

  if encrypted_data[0] == "!" # Apply CBC
    decipher = OpenSSL::Cipher.new("AES-256-CBC")
    decipher.iv = encrypted_data[1, 16] # Discard !
    encrypted_data = encrypted_data[17..-1]
  else # Apply ECB
    decipher = OpenSSL::Cipher.new("AES-256-ECB")
  end
  decipher.decrypt
  decipher.key = [key].pack "H*"

  begin
    return decipher.update(encrypted_data) + decipher.final
  rescue OpenSSL::Cipher::CipherError
    vprint_error "Vault password could not be decrypted with key #{key}"
    return nil
  end
end

def encrypt_vault_password_cbc(plaintext)
  iv = OpenSSL::Digest::MD5.digest("hello world")
  key = "A" * 32

  cipher = OpenSSL::Cipher.new("AES-256-CBC")
  cipher.encrypt
  cipher.iv = iv
  cipher.key = key

  encrypted_deets = ""
  encrypted_deets << cipher.update(plaintext)
  encrypted_deets << cipher.final
  encrypted_data = "!#{iv}#{encrypted_deets}"

  [key.unpack1("H*"), encrypted_data]
end

def encrypt_vault_password_ecb(plaintext)
  key = "A" * 32

  cipher = OpenSSL::Cipher.new("AES-256-ECB")
  cipher.encrypt
  cipher.key = key

  encrypted_deets = ""
  encrypted_deets << cipher.update(plaintext)
  encrypted_deets << cipher.final
  encrypted_data = "#{encrypted_deets}"

  [key.unpack1("H*"), encrypted_data]
end

key, encrypted_data = encrypt_vault_password_cbc("hunter2")
puts "Unencrypted cbc result: #{decrypt_vault_password(key, encrypted_data)}"

key, encrypted_data = encrypt_vault_password_ecb("hunter2")
puts "Unencrypted ebc result: #{decrypt_vault_password(key, encrypted_data)}"
```

</details>

File: modules/post/windows/gather/hashdump.rb [hunk 5]
```diff
@@ -145,15 +145,16 @@ def capture_hboot_key(bootkey)
         hash.update(vf[0x70, 16] + @sam_qwerty + bootkey + @sam_numeric)
 
         rc4 = OpenSSL::Cipher.new("rc4")
+        rc4.decrypt
         rc4.key = hash.digest
         hbootkey  = rc4.update(vf[0x80, 32])
         hbootkey << rc4.final
         hbootkey
```
Comment (adfoster-r7) [inline modules/post/windows/gather/hashdump.rb:152]: This works fine in Ruby 2.7 and 3.0, just updating for consistency

<details>

<summary>Code snippet</summary>

```ruby
require 'openssl'

value_data = "B" * 2000

hash = Digest::MD5.new
hash.update("hello world")

rc4 = OpenSSL::Cipher.new('rc4')
rc4.decrypt
rc4.key = hash.digest
hboot_key = rc4.update(value_data[0x80, 32])
hboot_key << rc4.final
hboot_key
```

</details>

File: modules/post/windows/gather/hashdump.rb [hunk 6]
```diff
@@ -145,15 +145,16 @@ def capture_hboot_key(bootkey)
...
       when 2
         aes = OpenSSL::Cipher.new('aes-128-cbc')
+        aes.decrypt
         aes.key = bootkey
         aes.padding = 0
-        aes.decrypt
         aes.iv = vf[0x78, 16]
         aes.update(vf[0x88, 16]) # we need only 16 bytes
```
Comment (adfoster-r7) [inline modules/post/windows/gather/hashdump.rb:159]: Example code snippet working in 2.7 but failing in 3:

```ruby
require 'openssl'

boot_key = "A" * 16
value_data = "B" * 2000

aes = OpenSSL::Cipher.new('aes-128-cbc')
aes.key = boot_key
aes.padding = 0
aes.decrypt
aes.iv = value_data[0x78, 16]
aes.update(value_data[0x88, 16])
```

Error:
```ruby
> aes.update(value_data[0x88, 16])
Traceback (most recent call last):
        5: from /usr/local/bin/irb:23:in `<main>'
        4: from /usr/local/bin/irb:23:in `load'
        3: from /usr/local/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        2: from (irb):17:in `<main>'
        1: from (irb):17:in `update'
OpenSSL::Cipher::CipherError (key not set)
```

After moving the decrypt call before the key is set it works as expected:

```ruby
require 'openssl'

boot_key = "A" * 16
value_data = "B" * 2000

aes = OpenSSL::Cipher.new('aes-128-cbc').decrypt
aes.decrypt
aes.key = boot_key
aes.padding = 0
aes.iv = value_data[0x78, 16]
aes.update(value_data[0x88, 16])
```

File: lib/msf/core/post/windows/priv.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/msf/core/post/windows/priv.rb:368]: I thought this was weird at the time having to reset this as part of the loop, as it goes against how `aes-256-cbc` is meant to function, i.e. the cipher blocks are chained together.

I'm thinking this is meant to be `aes-256-ecb` maybe? It passes the tests still:

```diff
-    aes = OpenSSL::Cipher.new("aes-256-cbc") 
+    aes = OpenSSL::Cipher.new("aes-256-ecb")
     aes.decrypt
+    aes.padding = 0
     aes.key = sha256x.digest

     vprint_status("digest #{sha256x.digest.unpack("H*")[0]}")

     decrypted_data = ''

     (60...policy_secret.length).step(16) do |i|
-      aes.reset
-      aes.padding = 0
       decrypted_data << aes.update(policy_secret[i,16])
     end

    return decrypted_data
```

**Comment by adfoster-r7** [comment]
Tested with some modules, all seems well

File: lib/msf/core/post/windows/priv.rb (hunk 2 above)
Comment (smcintyre-r7) [reply lib/msf/core/post/windows/priv.rb:368]: I can't seem to find any official Microsoft documentation to explain this. The way Metasploit does it is very similar to how impacket handles it, see [secretsdump.py:1081](https://github.com/SecureAuthCorp/impacket/blob/e1b9e247de5b4514ea3f5b0229e8187d9e8142d4/impacket/examples/secretsdump.py#L1081:9). Another implementation that I found is [RemoteHashRetrieval.ps1:943](https://github.com/HarmJ0y/DAMP/blob/master/RemoteHashRetrieval.ps1#L943) which does not appear to reset the cipher.

Comment (adfoster-r7) [reply lib/msf/core/post/windows/priv.rb:368]: I'm happy to leave this as-is just in case there's weird edge cases that aren't immediately obvious 👍

Comment (smcintyre-r7) [reply lib/msf/core/post/windows/priv.rb:368]: Yeah I spent a few hours looking through lsasrv.dll trying to find the exact encryption algorithm and was unsuccessful. Working in reverse, assuming that the CNG API is used, I only found references to CFB and CBC mode being used with AES encryption. I'm not entirely sure this is what's used though, I'd need to do more analysis.

image


Leaving it in its current state is probably our best bet.

Comment (adfoster-r7) [reply lib/msf/core/post/windows/priv.rb:368]: Thanks for taking a deeper look! 🎉

File: lib/rex/proto/ntlm/crypt.rb [hunk 7]
```diff
@@ -82,10 +82,11 @@ def self.des_56_to_64(ckey56s)
 
   def self.apply_des(plain, keys)
     raise RuntimeError, "No OpenSSL support" if not @@loaded_openssl
-    dec = OpenSSL::Cipher::DES.new
+    dec = OpenSSL::Cipher.new('DES')
     keys.map do |k|
+      dec.decrypt
       dec.key = k
-      dec.encrypt.update(plain)
+      dec.update(plain)
```
Comment (smcintyre-r7) [inline lib/rex/proto/ntlm/crypt.rb:89]: This appears to flip the usage from encrypt to decrypt here. Based on the input being "plain", I think it means to encrypt the data.

**Review by smcintyre-r7 — COMMENTED** [review]
Everything makes sense and looks good to me except for the one comment I left which I'll look at testing / reviewing more thoroughly.

File: lib/rex/proto/ntlm/crypt.rb (hunk 7 above)
Comment (smcintyre-r7) [reply lib/rex/proto/ntlm/crypt.rb:89]: I'm pretty certain this whole block was broken prior to this and is still broken.

TLDR; this should be:
```suggestion
      dec.encrypt
      dec.key = k
      dec.update(plain) + dec.final
```

First, on inspecting `#lm_hash` form IRB, it's not generating a hash.

```
>> Rex::Proto::NTLM::Crypt.lm_hash('test')
=> ""
```

Now, I'm pretty confident that `#apply_des` is returning an array of two empty strings.
```
>> keys = Rex::Proto::NTLM::Crypt.gen_keys('TEST'.ljust(14, "\0"))
=> ["T\"Tj@\x00\x00\x00", "\x00\x00\x00\x00\x00\x00\x00\x00"]
>> Rex::Proto::NTLM::Crypt.apply_des(Rex::Proto::NTLM::Crypt::CONST::LM_MAGIC, keys)
=> ["", ""]
```

This is probably because `dec.update` returns an empty string. That's because after the cipher (en|de)crypt operation, the `#final` method needs to be called, making this `dec.update(plain) + dec.final`. Upon testing this however, I receive a ```in `final': bad decrypt (OpenSSL::Cipher::CipherError)``` exception. After changing it back to `encrypt` though, I see the value I would [expect](https://asecuritysite.com/encryption/lmhash).

```
>> Rex::Proto::NTLM::Crypt.lm_hash('test')
=> "\x01\xFCZk\xE7\xBCi)\x03\xBE\xC6\xE2\xE0\xDCX\f\xAA\xD3\xB45\xB5\x14\x04\xEEG\xD6\x84\xD8\xC0\xF0\xDE\xA9"
```

Comment (adfoster-r7) [reply lib/rex/proto/ntlm/crypt.rb:89]: Thanks for catching this! I took a look at this code and started writing tests for it. I then came to the realisation that the code actually exists in [rubyntlm](https://github.com/WinRb/rubyntlm) too. I'm not sure what the historical context is here, or which came first 😄 

Interestingly I also spotted the same fix that you're suggesting in the rubyntlm library too:
https://github.com/WinRb/rubyntlm/commit/f04cff7c6499217b99f3d002f90c1a3aaec25969

As a sanity check, I pointed the rubyntlm's [tests](https://github.com/WinRb/rubyntlm/blob/59879617d14183bca538696722844dc582b4101f/spec/lib/net/ntlm_spec.rb) against crypt.rb and it spotted a different bug with the current implementation. This function doesn't appear to exist on `Rex::Proto::NTLM::Base`, but we do have a Rex::Text equivalent
https://github.com/rapid7/metasploit-framework/blob/19bc85fa1d418c90bf43f8f203f39d33bfdd2430/lib/rex/proto/ntlm/crypt.rb#L132

Our `gen_keys` implementation is also quite different too, and currently fails with rubyntlm's specs

I can put up a separate pr to fix all those issues and potentially reuse the same tests from the ntlm library to verify its behavior, or I could simply keep the same API and delegate responsibility to that gem - as we're already including it:
https://github.com/rapid7/metasploit-framework/blob/19bc85fa1d418c90bf43f8f203f39d33bfdd2430/metasploit-framework.gemspec#L80-L81

Any strong preference? 😄

Comment (smcintyre-r7) [reply lib/rex/proto/ntlm/crypt.rb:89]: Not a strong preference but delegating the responsibility to reduce duplicate code sounds like a good option.

**Comment by smcintyre-r7** [comment]
Is there anything I can do to help advance this?

**Comment by adfoster-r7** [comment]
@smcintyre-r7 I've cherry picked this PR and your suggested fix into a Rails6/Ruby 3 update PR and I've been testing that. My latest thoughts were to test and land that all together, as we'll be doing a thorough test of that branch before it's landed. 

Alternatively, I'm more than happy to make your suggested tweak and get this landed in isolation if you think it's relatively low risk/test effort! 👍 

Will be punting on using the third party rubyntlm gem either way though, as I think it's too much scope creep from the original isolated goal upgrading to Ruby 3

**Comment by adfoster-r7** [comment]
Closing in favor of a final ruby 3.x PR now that Rails 6.1 support is merged :+1:
